### PR TITLE
Issue #2574 - Improve ability to resume from failed reindex

### PR DESCRIPTION
**Pull request by tbieste**
Issue #2574 - Improve ability to resume from failed reindex

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java [hunk 1]
```diff
@@ -298,30 +370,36 @@ private boolean extractIndexIds(Parameters output) {
...
+                        logger.severe("IllegalStateException caught. FHIR client thread will exit." + x.toString() );
+                    }
+                    if (!ok) {
+                        // stop everything on the first failure
+                        this.active = false;
+                    }
                 }
             }
+        } catch (Throwable t) {
+            logger.severe("Throwable caught. FHIR client thread will exit." + t.toString() );
```
Comment (lmsurpre) [inline fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java:397]: does Throwable.toString print the stacktrace?
maybe better to do this?
```suggestion
            logger.log("Throwable caught. FHIR client thread will exit.", t);
```

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ServerDrivenReindexOperation.java [hunk 2]
```diff
@@ -190,27 +190,32 @@ public void waitForStop() {
...
+                    // So probably this will be caught first time monitorLoop calls callOnce and not here.
+                    logger.severe("IllegalStateException caught. FHIR client thread will exit." + x.toString() );
+                }
+                if (!ok) {
+                    // stop everything on the first failure
+                    this.active = false;
+                }
             }
+        } catch (Throwable t) {
+            logger.severe("Throwable caught. FHIR client thread will exit." + t.toString() );
```
Comment (lmsurpre) [inline fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ServerDrivenReindexOperation.java:214]: same as above
```suggestion
            logger.log(Level.SEVERE, "Throwable caught. FHIR client thread will exit.", t);
```

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java [hunk 3]
```diff
@@ -298,30 +370,36 @@ private boolean extractIndexIds(Parameters output) {
...
-                    // stop everything on the first failure
-                    this.active = false;
+        try {
+            while (this.running && this.active) {
+                String indexIds = getIndexIdsToReindex();
+                if (!indexIds.isEmpty()) {
+                    boolean ok = false;
+                    try {
+                        ok = callReindexOperation(indexIds);
+                    } catch (DataAccessException x) {
```
Comment (JohnTimm) [inline fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java:380]: Should we just catch Throwable here and treat any and all exceptional cases the same? This way we don't need these two specific catches, we need only set active = false once, and we don't need the outer try-catch.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java (hunk 1 above)
Comment (tbieste) [reply fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java]: Updated.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ServerDrivenReindexOperation.java (hunk 2 above)
Comment (tbieste) [reply fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ServerDrivenReindexOperation.java]: Updated.

**Review by lmsurpre — APPROVED** [review]
I checked out the branch and ran it against a large (70+ million resource) db.
When the server went offline this morning, the log output looked better and eventually the tool exited with a message like "Reindexing was not fully completed, restart reindex from index ID '19611009' to resume".  LGTM.

File: fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java (hunk 3 above)
Comment (JohnTimm) [reply fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java]: Created and merged separate PR with suggested changes.

**Review by JohnTimm — APPROVED** [review]
LGTM

**Pull request merged**